README
                             dist 4.0
                       **  Alpha version **

                   Copyright (c) 1988, Larry Wall
                Copyright (c) 1990-1992, Harlan Stenn
              Copyright (c) 1991-1997, Raphael Manfredi
              Copyright (c) 2004-2006, Raphael Manfredi

------------------------------------------------------------------------
    This program is free software; you can redistribute it and/or modify
    it under the terms of the Artistic License, a copy of which can be
    found with this package.

    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    Artistic License for more details.
------------------------------------------------------------------------

This version of dist requires you to have at least perl 4.0 PL36.
It has also been ported to work with perl 5.0 PL0, provided you have
at least integrated patches 0a-0h, issued by Andy Dougherty and made
available at the following ftp sites:

	ftp.demon.co.uk:/pub/perl/db/perl5.000-p0a-p0h.tar.gz
	ftp.funet.fi:/pub/languages/perl/ports/perl5/perl5.000-p0a-p0h.tar.gz

Please read all the directions below before you proceed any further, and
then follow them carefully.

After you have unpacked your kit, you should have all the files listed
in MANIFEST.

========================================================================
                           It's not the way I wrote it, but the
                           way you play it makes it sound a lot better.
                                -- Camille Saint-Saens (French composer)

The dist package consists of four parts:

    1) The Configure generator (metaconfig) and its supporting files.
    2) The distribution kit maker (makedist) and its supporting files.
    3) The patch distribution system (pat*) and its supporting files.
    4) The Makefile generator (jmake) and its supporting files.

Each of these can be used separately from the others.  Before you can
use any of those, however, the directory containing your package (not this
package) must be initialized by putting a .package file into it.  This
can be done by running packinit in that directory.

There is a mailing list hosted in Japan and set up by Shigeya Suzuki
<shigeya@foretune.co.jp>, for discussion about the dist package as a
whole. It's a good place to ask questions (or answer them) and to
send your patches. I will post official patches to the net, as well
as to the dist-users list.

To send a mail to the list, address it to <dist-users@foretune.co.jp>.
To subscribe, send a mail to <majordomo@foretune.co.jp>. If you don't
know how to use majordomo, the syntax of the subscribe command is:

	subscribe dist-users [address]

where the address part is optional. You may unsubscribe automatically
at any time by sending:

	unsubscribe dist-users

If you have a problem with this version of dist, it is recommended that
you subscribe to the list, then send a description of your problem to it.
If you send mail to me personally, I may not be able to answer in a
timely fashion.

This mailing list has low traffic (a few articles per week, typically),
and it is expected to remain so, with a high signal/noise ratio.

Notes:

    If you are running on a system with <= 14 char filenames, don't
    distribute any files with filenames longer than 12 chars (with the
    exception of patchlevel.h), so that there will be room for a
    2-digit extension indicating patch number in your bugs direcory.
    This includes .SH files, so any shell script built by a .SH file
    should be 9 characters or less.  On systems with flexfilenames you
    don't have to worry about it.

    This package has been designed on top of dist 2.0, which has been
    written by Larry Wall <lwall@netlabs.com>.
========================================================================

INSTALLATION

1) Run Configure. This will figure out various things about your
system. After it has completed, it will produce config.h and config.sh.

You might possibly have to trim # comments from the front of Configure
if your shell doesn't handle them, but all other comments will be taken
care of.

2) Run make.

3) If make succeeded, you may wish to do "make install install.man". Be
sure your rights are correct (if you install manual pages, you may need
super-user privileges). By not running "make install.man", you avoid the
installation of the manual pages.

4) Read the manual entry before running.

5) IMPORTANT!  Communicate any problem and suggested patches to me,
ram@hptnos02.grenoble.hp.com (Raphael Manfredi), so we can keep this
distribution in sync.  If you have a problem, there will be someone else
who had it or will have it too...

If possible, send me patches such that the patch program will apply
them.  Context diffs are the best, then normal diffs.  Do not send ed
scripts, I have probably changed my copy since the version you got.

6) After everything is installed, you can do make clobber. This will
clean up everything and let you re-distribute this kit, without
carrying useless files. You should keep this distribution intact, so
that future patches will be applyable.

7) I have an automatic patch sender. Send me the following mail:

	Subject: Command
	@SH mailhelp PATH

and you'll get instructions (PATH stands for YOUR e-mail address, either
in INTERNET or in bang notation). I would recommend you to get all the
issued patches before you start making some modifications on this
package.

8) If you wish to deinstall the package, you may run "make deinstall".
A separate "make deinstall.man" will remove the manual pages. Be sure
the makefiles are correctly set before running any deinstall target.
On USG systems, some executable have a chance to remain despite the
deinstall (text file busy...).

	Raphael Manfredi <Raphael_Manfredi@grenoble.hp.com>


ChangeLog
Fri Feb 28 17:48:41 MET 1997   Raphael Manfredi <Raphael_Manfredi@grenoble.hp.com>

. Description:

	This huge set of patch results from an integration of the
	perl5 units into the dist-3.0 package. A few improvements
	of mine are also incorporated here, but this patch set should
	mainly be credited to Andy Dougherty and Chip Salzenberg, who
	have both assumed the perl5 Patch Pumpkin with success and brio.

	I won't document most of the changes in detail. Only the obvious
	ones were kept here. You can look closer to the patches themselves
	to see what really changed. Especially the new units, since you
	now have more symbols to check for in your C code...

	Regenerated Configure with current set of units.

	Added Perl Patch Pumpkin credits. Thanks to Andy Dougherty and
	to Chip Salzenberg for their incredible contributions to
	metaconfig.

	Changed my email address and updated copyright dates

	Added support for src.U. This means you can now run Configure
	from a remote directory to separate the build directory and
	the (possibly read-only) source tree.
	Note that some of your private units may require some changes
	due to that...

	Removed support for NO_PROTOTYPE detection on SCO.

	Added cute quoting trick for wild stringify support, which
	allows non-ANSI stringification to occur.

	Replaced .a with $_a all over the place.
	Likewise for .o replaced by $_o.
	You'll have to do that to your own private units.

	Integrated new units from perl5. I will not document most
	of the perl5-related changes here. People interested can
	look at Andy's notes in the perl5 distribution.

	New USE_BSD_GETPGRP to detect the getpgrp() flavour. The old
	USE_BSDPGRP has been obsoleted in favor of USE_BSD_SETPGRP.
	Indeed, setpgrp() and getpgrp() can have distinct flavors,
	so we need two distinct tests for them.

	Useless units dropped: d_group.U and d_passwd.U.

	Mention that <poll.h> can be included if HAS_POLL defined.

	Improved overlapping copy check.

	Added support for Free_t, the type of free().

	Don't use nm with the GNU C library.
	Added support for Linux shared libs.

	Don't prompt them if the void support is high enough for us
	and don't ask for the include path unless they are on a MIPS.

	Added the srcdir support to "configure" now that we have src.U.

	Documented the runnning environment and the src.U unit in
	the metaconfig man page. The $_a and $_o are also mentionned.
	All the generated scripts must now start with a "$startsh"
	to ensure proper shell execution.

	New "create" and "empty" lint directives.

	New -L option for metaxref to match metaconfig and metalint.

	Documents contents of the message sent by patnotify:
	we now let them know the patch priority and description.

	Added a whole bunch of new units, mostly from the perl5 team.

. Files changed:

	* bin/packinit.SH: Fixed one wrong ':' comment in .package.

	* jmake/files/Jmake.rules: Now handles USRINC for dependencies. 
	Smarter about dependencies computation.

	* jmake/files/Jmake.tmpl: Lex path can now be configured.

	* mcon/U/Begin.U: Added Extractall dependency.

	* mcon/U/Config_sh.U, mcon/U/Oldsym.U, mcon/U/Extract.U,
	mcon/U/patchlevel.U, mcon/U/MailAuthor.U, mcon/U/Chk_MANI.U,
	mcon/U/Config_h.U: Added support for src.U.

	* mcon/U/Configdir.U: Have README explicitly mention the package
	name.

	* mcon/U/make.U, mcon/U/d_strtoul.U, mcon/U/Unix.U,
	mcon/U/d_inetaton.U, mcon/U/sitearch.U, mcon/U/d_sanemcmp.U,
	mcon/U/i_locale.U, mcon/U/d_sigaction.U, mcon/U/d_getpgid.U,
	mcon/U/src.U, mcon/U/sitelib.U, mcon/U/d_brokstat.U, mcon/U/man3dir.U,
	, mcon/U/d_sigsetjmp.U, mcon/U/i_sysstat.U, mcon/U/d_strtod.U,
	mcon/U/d_byacc.U, mcon/U/sh.U, mcon/U/Signal.U, mcon/U/d_gnulibc.U,
	mcon/U/man1dir.U, mcon/U/Extractall.U, mcon/U/i_values.U,
	mcon/U/i_sfio.U: Created.

	* mcon/U/Getfile.U: Getfile script now begins with "startsh".

	* mcon/U/Head.U: Make sure we unset CDPATH for shells that support
	this.  Improved Korn shell detection and handling.

	* mcon/U/Instruct.U: Logname / whoami sequence rewritten to use case.

	* mcon/U/Loc.U: Allow users to specify paths on the command line. 
	Will now substitute cp for ln if not supported.

	* mcon/U/Myread.U: Myread script now starts with a "startsh". 
	Miscellaneous fixes.

	* mcon/U/Oldconfig.U: Added support for src.U.  New OSNAME define. 
	Can now sense new OSes.

	* mcon/U/Options.U: Optdef.sh now starts with a "startsh".  Moved
	some code from Head.U.

	* mcon/U/Whoa.U: Whoa script now starts with leading "startsh".

	* mcon/U/afs.U: Can now explicitly tell Configure whether AFS is
	running.

	* mcon/U/archlib.U: Skip existence checks for archlib.

	* mcon/U/archname.U: Changed the way the archname is mangled from
	uname.

	* mcon/U/byteorder.U: No longer ask the user if the test runs ok.

	* mcon/U/ccflags.U: Removed support for NO_PROTOTYPE detection on
	SCO.  New locincpth variable.  Added info on the "additional ld
	flags" question.

	* mcon/U/cf_who.U: New computation method avoiding use of temporary
	file.

	* mcon/U/cpp_stuff.U: Added cute quoting trick for wild stringify
	support.

	* mcon/U/d_bsdjmp.U, mcon/U/d_fd_set.U, mcon/U/models.U,
	mcon/U/d_scannl.U, mcon/U/d_open3.U, mcon/U/d_closedir.U: Added ?F:
	metalint hint.

	* mcon/U/libs.U, mcon/U/libnm.U, mcon/U/libyacc.U, mcon/U/libflex.U,
	mcon/U/libdbm.U, mcon/U/d_socket.U, mcon/U/d_normsig.U,
	mcon/U/d_crypt.U: Replaced .a with $_a all over the place.

	* mcon/U/d_csh.U: Added full_csh to preserve the full path even when
	portable.

	* mcon/U/d_dosuid.U: Moved unit to TOP via a ?Y: layout directive. 
	Tell them /dev/fd is not about floppy disks.

	* mcon/U/d_gconvert.U: Integrated new unit from perl5.

	* mcon/U/d_getpgrp.U: New USE_BSD_GETPGRP to detect the getpgrp()
	flavour.

	* mcon/U/d_group.U, mcon/U/d_passwd.U: Useless unit dropped.

	* mcon/U/d_usendir.U, mcon/U/d_havetlib.U, mcon/U/errnolist.U:
	Replaced .a with $_a all over the place.  Likewise for .o replaced by
	$_o.


pat/pat.SH
case $CONFIG in
'')
	if test -f config.sh; then TOP=.;
	elif test -f ../config.sh; then TOP=..;
	elif test -f ../../config.sh; then TOP=../..;
	elif test -f ../../../config.sh; then TOP=../../..;
	elif test -f ../../../../config.sh; then TOP=../../../..;
	else
		echo "Can't find config.sh."; exit 1
	fi
	. $TOP/config.sh
	;;
esac
case "$0" in
*/*) cd `expr X$0 : 'X\(.*\)/'` ;;
esac
echo "Extracting pat/pat (with variable substitutions)"
cat >pat <<!GROK!THIS!
$startperl
	eval 'exec perl -S \$0 \${1+"\$@"}'
		if \$running_under_some_shell;

# $Id$
#
#  Copyright (c) 1991-1997, 2004-2006, Raphael Manfredi
#  
#  You may redistribute only under the terms of the Artistic Licence,
#  as specified in the README file that comes with the distribution.
#  You may reuse parts of this distribution only within the terms of
#  that same Artistic Licence; a copy of which may be found at the root
#  of the source tree for dist 4.0.
#
# Original Author: Larry Wall <lwall@netlabs.com>
#
# $Log: pat.SH,v $
# Revision 3.0.1.5  1994/10/29  16:37:53  ram
# patch36: now unlinks all the files created by patlog in bugs
#
# Revision 3.0.1.4  1994/01/24  14:29:17  ram
# patch16: added ~/.dist_profile awareness
#
# Revision 3.0.1.3  1993/08/25  14:04:35  ram
# patch6: removal of patch temporary files did not work with gaps
#
# Revision 3.0.1.2  1993/08/24  12:14:39  ram
# patch3: now removes older patch temporary files within bugs
#
# Revision 3.0.1.1  1993/08/19  06:42:31  ram
# patch1: leading config.sh searching was not aborting properly
#
# Revision 3.0  1993/08/18  12:10:36  ram
# Baseline for dist 3.0 netwide release.
#

\$version = '$VERSION';
\$patchlevel = '$PATCHLEVEL';
!GROK!THIS!
cat >>pat <<'!NO!SUBS!'

$progname = &profile;				# Read ~/.dist_profile
require 'getopts.pl';
&usage unless $#ARGV >= 0;
&usage unless &Getopts("ahmnV");

if ($opt_V) {
	print STDERR "$progname $version PL$patchlevel\n";
	exit 0;
} elsif ($opt_h) {
	&usage;
}

&readpackage;

if (-f 'patchlevel.h') {
	open(PL,"patchlevel.h") || die "$progname: can't open patchlevel.h: $!\n";
	while (<PL>) {
		$bnum = $1 if /^#define\s+PATCHLEVEL\s+(\d+)/;
	}
	die "$progname: malformed patchlevel.h file.\n" if $bnum eq '';
	++$bnum;
} else {
	$bnum=1;
}

if ($opt_n) {
	&newer;				# Look for files newer than patchlevel.h
} elsif ($opt_a) {
	open(MANI,"MANIFEST.new") || die "No MANIFEST.new found.\n";
	@ARGV = ();
	while (<MANI>) {
		s|^\./||;
		next if m|^patchlevel.h|;		# This file is built by hand
		chop;
		($_) = split(' ');
		next if -d;
		push(@ARGV,$_);
	}
	close MANI;
} elsif ($opt_m) {
	open(MODS,"bugs/.mods$bnum") || die "$progname: no modification found.\n";
	@ARGV = ();
	while (<MODS>) {
		next if m|^patchlevel.h$|;		# This file is built by hand
		chop;
		($_) = split(' ');
		push(@ARGV,$_);
	}
	close MODS;
}

# Remove older patch temporary files

@patlist = &patseq($bnum - 1);
foreach $cpat (@patlist) {
	unlink <bugs/*.$cpat bugs/.logs$cpat bugs/.mods$cpat bugs/.xlog$cpat>;
	unlink <bugs/.pri$cpat bugs/.subj$cpat bugs/.clog$cpat bugs/.rlog$cpat>;
}

# Since we're about to launch other pat exectuables, disable ~/.dist_profile
# to protect them if they added some weird switches we don't need...

$ENV{'DIST'} = '/dev/null';		# Disable ~/.dist_profile

system 'perl', '-S', 'patcil', '-p', @ARGV;

# Update MANIFEST if necessary, then patcil it.

if (-f 'MANIFEST' && `diff MANIFEST.new MANIFEST 2>/dev/null` ne '') {
	system 'cp', 'MANIFEST.new', 'MANIFEST';
	system 'perl', '-S', 'patcil', '-p', 'MANIFEST';
	push(@ARGV, 'MANIFEST');
}

system 'perl', '-S', 'patdiff', @ARGV;
system 'perl', '-S', 'patmake';

sub usage {
	print STDERR "Usage: $progname [-ahmnV] [filelist]\n";
	print STDERR "  -a : all the files in MANIFEST.new\n";
	print STDERR "  -h : print this message and exit\n";
	print STDERR "  -m : all the modified files (which have been patciled)\n";
	print STDERR "  -n : all the files newer than patchlevel.h\n";
	print STDERR "  -V : print version number and exit\n";
	exit 1;
}

!NO!SUBS!
$grep -v '^;#' ../pl/newer.pl >>pat
$grep -v '^;#' ../pl/package.pl >>pat
$grep -v '^;#' ../pl/patseq.pl >>pat
$grep -v '^;#' ../pl/tilde.pl >>pat
$grep -v '^;#' ../pl/profile.pl >>pat
chmod +x pat
$eunicefix pat
